File: src/satisfactory_mcp/interfaces/web/frontend/src/placements.ts
/* What the player built and where it stands: the floor plan, the machines on it, and the
 * containers among them.
 *
 * Together because they are the same drawing problem -- a rotated footprint at a measured
 * size -- and because all three fall back to a stated stand-in when the docs dump carries no
 * clearance for a class. The difference is only where the size comes from: a foundation
 * takes the grid's own tile, a machine and a container take their own w_m and l_m.
 *
 * The storage layer at the bottom is the one that is not only a placement: a container's
 * CONTENTS are the point of drawing it, so it is the one thing here with a popup built from
 * data rather than from a fixed set of rows.
 */

import { CONTENTS_POPUP_PX, code, contentsRows, count, popup } from "./dom";
import type { Row } from "./dom";
import { refreshFloors } from "./floors";
import { L } from "./leaflet";
import { BAND, layer } from "./layers";
import { footprintCorners } from "./map";
import { raiseNodeDots } from "./markers";
import { declareColours } from "./palette";
import { registerFetch } from "./registry";

import type {
  MachinesResponse,
  StorageResponse,
  StorageRow,
  StructuresResponse,
} from "./api-types";

/* The player's floor plan: one 8 m tile per placed foundation, ramp, wall or catwalk.
 *
 *   * Drawn at its real yaw, as of schema 12. Until then the instance quaternion was
 *     dropped at extraction and every tile was axis-aligned, so a slab the player laid at
 *     an angle -- this world has several, over 1,800 pieces of them -- came out as a
 *     staircase of squares. Polygons rather than rectangles is the whole change: an
 *     L.rectangle IS a polygon built from two corners, so this costs four points per piece
 *     and nothing else.
 *   * No per-class size. None of these eighteen classes has clearance data, so there is
 *     no footprint to ask for. They all snap to the same grid, whose edge the server
 *     reports as `tile_m`, so a wall paints the tile it stands on rather than its own
 *     thin volume -- it straddles two tiles and fringes a walled platform by half a tile,
 *     which at any zoom where the platform is legible is not visible.
 *
 * Stroked in its own fill colour, the trick the biome cells already use: no stroke at all
 * leaves hairline seams between neighbouring tiles at low zoom, and a stroke in any other
 * colour draws an 8 m grid. Same colour, weight 1, and a slab reads as one platform.
 */
/* Concrete, meant to be cool enough to read as built against the biomes -- and that sentence
 * is the one claim in this codebase the executable version of the discipline contradicts
 * outright: it is dE 6.6 from Abyss Cliffs and under 15 from three more grounds. Listed in
 * STANDING in palette.ts rather than quietly fixed, because which of the two colours should
 * move is a decision to make against the map. */
var STRUCTURE_COLOUR = declareColours("placements", { foundations: "#3a4148" }).foundations;

export function drawStructures(data: StructuresResponse): void {
  // First of the built band, because the concrete is what everything else in it stands on
  // or runs over -- the legend reads a base bottom-up, exactly as the player laid it.
  var group = layer("foundations", true, STRUCTURE_COLOUR, [BAND.built, 0, "foundations"]);
  // No `|| 8`: `tile_m` is the server's FOUNDATION_M constant and is always sent, and the
  // fallback was a second copy of the number the field exists to stop the page hardcoding.
  var half = data.tile_m / 2;
  data.structures.forEach(function (s, row) {
    if (s.x_m === null || s.y_m === null) return;
    var piece = L.polygon(footprintCorners(s.x_m, s.y_m, half, half, s.yaw), {
      color: STRUCTURE_COLOUR,
      weight: 1,
      opacity: 0.9,
      fillColor: STRUCTURE_COLOUR,
      fillOpacity: 0.9,
      interactive: false,
      pane: "foundations",
    });
    // The only name a lightweight buildable has is its place in this list, so that is what
    // the floor filter joins on -- `deck_rows` in `/api/floors` indexes exactly this. The
    // index is taken from the payload rather than from a counter, so a row skipped by the
    // guard above does not shift every piece after it by one. See floors.ts.
    piece._floor = { row: row, x_m: s.x_m, y_m: s.y_m, z_m: s.z_m === null ? undefined : s.z_m };
    piece.addTo(group);
  });
}

/* Static, and early in the wave: the concrete is what floor mode measures a platform's extent
 * from, so the opening flight cannot happen until these pieces are on the map. */
registerFetch<StructuresResponse>({
  wave: "static",
  rank: 20,
  path: "/api/structures",
  label: "structures",
  clears: ["foundations"],
  refilters: true,
  draw: drawStructures,
});

/* Machines at their real size AND their real facing: `w_m`/`l_m` are the building's own
 * footprint, so a Manufacturer (18x20 m) reads as the eight-times-larger thing it is next
 * to a Constructor (8x10 m), and `yaw` turns that footprint the way the player placed it.
 * Null footprints for the classes the docs dump gives no clearance data -- both biomass
 * burners here -- and those fall back to the 6 m square every machine used to get, which
 * rotates to itself. A null yaw draws axis-aligned; see footprintCorners in map.ts for why that is
 * not the same statement as a yaw of zero. */
var MACHINE_FALLBACK_M = 6;

/* Blue, amber, red -- the page's oldest three colours, and the ones every measured warrant
 * since has had to get out of the way of: the pipe rust was chosen against this amber, the
 * storage magenta against this red, and the wire violet against all three. Not themselves
 * measured against anything, which is what the audit has to say about the water node dot: a
 * water extractor stands exactly on one, and the blue is dE 8.6 away. */
var KIND_COLOUR: Record<string, string> = declareColours("placements", {
  machines: "#4aa3df",
  extractors: "#e0a33f",
  generators: "#d9534f",
});

/* The three layers /api/machines answers with, and the one row shape all three carry.
 * Spelled as a tuple rather than inferred, so `data[kind]` is a PlacementRow[] rather than
 * an index into an object with a string. */
const MACHINE_KINDS = ["machines", "extractors", "generators"] as const;

/* Where each of the three sits in the built band, spelled rather than taken from the loop's
 * own index: the order these are FETCHED and drawn in is one payload's field order, and the
 * order they are LISTED in is an editorial choice about a legend. They agree today. Reading
 * the second off the first would make the day they stop agreeing a silent one. */
const MACHINE_SLOT: Record<(typeof MACHINE_KINDS)[number], number> = {
  machines: 40,
  extractors: 50,
  generators: 60,
};

export function drawMachines(data: MachinesResponse): void {
  MACHINE_KINDS.forEach(function (kind) {
    var group = layer(kind, kind !== "machines", KIND_COLOUR[kind], [
      BAND.built,
      MACHINE_SLOT[kind],
      kind,
    ]);
    data[kind].forEach(function (m) {
      // One guard, on x only, exactly as it always was. The row type says y_m can be null
      // too and the assertion below is that claim being deliberately not acted on: widening
      // this to `|| m.y_m === null` would be a behaviour change smuggled in as a type fix,
      // and if the projection ever sends half a position it should be visible, not silently
      // skipped by a guard nobody decided to add.
      if (m.x_m === null) return;
      var w = (m.w_m || MACHINE_FALLBACK_M) / 2;
      var l = (m.l_m || MACHINE_FALLBACK_M) / 2;
      var piece = L.polygon(footprintCorners(m.x_m, m.y_m!, w, l, m.yaw), {
        color: KIND_COLOUR[kind],
        weight: 1,
        fillOpacity: m.paused ? 0.15 : 0.65,
        dashArray: m.paused ? "2,2" : undefined,
      }).bindPopup(
        popup([
          ["building", m.name],
          ["recipe", m.recipe_name || m.recipe],
          ["clock", m.clock === null ? null : Math.round(m.clock * 100) + "%"],
          ["paused", m.paused ? "yes" : null],
          // All three sides of the clearance box, because the third is now sent: a
          // Refinery being 15 m tall is why a floor view can say it comes through the
          // ceiling, and a reader looking at the ghost should find the number here.
          ["footprint", m.w_m && m.l_m ? m.w_m + " x " + m.l_m + " m" : null],
          ["height", m.h_m ? m.h_m + " m" : null],
          // Degrees about world Z, positive turning +X towards +Y -- the same number
          // the drawing is turned by, so a reader can check the picture against it.
          // Absent, not "0", when the projection carries no facing at all.
          ["facing", m.yaw === null || m.yaw === undefined ? null : Math.round(m.yaw) + "°"],
          ["at", m.x_m + ", " + m.y_m + " m"],
          ["instance", code(m.instance_leaf)],
        ])
      );
      // What the floor filter joins a machine by, and what it needs to know to tell whether
      // one on a lower deck comes up through this floor. See floors.ts.
      piece._floor = {
        id: m.instance_leaf,
        z_m: m.z_m === null ? undefined : m.z_m,
        h_m: m.h_m,
      };
      piece.addTo(group);
    });
  });
  raiseNodeDots();
}

/* The live wave's first entry, and the only one on the page with a post-draw hook.
 *
 * A save write changes what is BUILT, so it changes the decomposition -- and the ids a band
 * lists are what the floor filter runs on. Without `after` a machine placed since the view
 * was opened would be drawn by /api/machines, listed by no band, and therefore silently
 * missing from every floor rather than visibly new on one. It is a hook rather than a line at
 * the end of `drawMachines` because it is a refetch and not a draw: it asks /api/floors the
 * same question again, which is load.ts's kind of work, not this file's. */
registerFetch<MachinesResponse>({
  wave: "live",
  rank: 10,
  path: "/api/machines",
  label: "machines",
  clears: ["machines", "extractors", "generators"],
  refilters: true,
  draw: drawMachines,
  after: refreshFloors,
});


/* Storage: the boxes, and what is in them.
 *
 * Here rather than in a file of its own because it is the same drawing problem this module
 * already solves twice -- a rotated footprint at a measured size, with a stated stand-in where
 * the dump has no clearance -- and a third copy of `footprintCorners` plus a third fallback
 * constant would be three places to keep in step instead of one. What is genuinely new is the
 * popup, because a container is the first thing on this map whose CONTENTS are the point.
 *
 * The map could already say what the player owns -- the header's totals come from the same
 * stacks -- and could never say where any of it was. That is the gap: "have I got enough
 * steel" and "where did I put the steel" are different questions, and a base with 105
 * containers spread over 7 km only ever raises the second one.
 *
 * ONE LAYER, TWO KINDS, told apart by a value step in one hue -- see STORAGE_COLOUR. A fluid
 * buffer and a storage container are both boxes the player put things in, so they belong to one
 * checkbox; they hold different sorts of thing and read differently in a popup, so they are not
 * the same tone. That is the belts' and pipes' grammar borrowed for a distinction that is not a
 * tier, and it is deliberate: a second hue would make the legend claim these are two networks.
 *
 * OFF BY DEFAULT, and NOT part of the reveal a factory label triggers -- see FACTORY_LAYERS in
 * labels.ts for that decision and its reasoning.
 */

/* Storage, and picked the way the pipe rust was: by measuring, not by taste.
 *
 * A container is drawn as a filled footprint box, so the colours it has to separate from are
 * the other filled boxes -- the three machine kinds, the belt attachments, and the concrete it
 * stands on -- and then, more weakly, everything else on the page. Magenta is what is left: the
 * page already spends blue on machines, amber on extractors, red on generators, steel on belts
 * and rust on pipes, and the whole warm half is taken.
 *
 * In CIE Lab, this is dE 51.8 from its nearest filled box (the generator red) and 48.5 from
 * the nearest biome ground, which are the two comparisons that decide whether a box reads. Its
 * nearest neighbour ANYWHERE on the page is the raw-quartz node dot at dE 27.4 -- a small disc
 * on open terrain rather than a rectangle inside a factory, so the two are never asked to be
 * told apart in the same square metre. That last sentence is the axis palette.ts's STANDING
 * list is organised along, and this is the one place on the page where it is backed by a
 * measurement rather than assumed. The alternatives measured beside it were all worse on one
 * of the two: a lighter magenta (#c76bb0) lands dE 17.7 from that same quartz dot, a violet
 * (#8c72c4) dE 19.1 from the crude-oil dot and only 37.3 from the machine blue, and a sea
 * green dE 10.5 from the pickup teal.
 *
 * And the fluid buffers, one value step down the same hue -- the grammar the belts and pipes
 * use for their tiers, borrowed for a distinction that is not a tier: a tank and a box are two
 * kinds of container rather than two grades of one, and one family with a step inside it says
 * "same layer, different thing" without spending a second hue on it.
 *
 * The step is the house step: dE 16.7, against the belts' 15.6 between their slowest and
 * fastest and the pipes' 15.7 between Mk1 and Mk2. Re-measured rather than assumed safe,
 * because a ramp can walk a colour into a neighbour -- this one moves AWAY from everything,
 * ending dE 34.0 from its nearest colour on the page (the crude-oil dot) and 37.6 from the
 * nearest ground, both further off than the box tone above. Both steps are inside one owner,
 * which is exactly why palette.ts compares across owners and never within one.
 */
var STORAGE = declareColours("placements", {
  storage: "#ad4f96",
  "storage fluid": "#7f3169",
});
var STORAGE_COLOUR = STORAGE.storage;
var STORAGE_FLUID_COLOUR = STORAGE["storage fluid"];

/* A container the docs dump carries no clearance for: the HUB's own box, the Blueprint
 * Designer's, and the Dimensional Depot uploader. The same arrangement, and the same reason, as
 * MACHINE_FALLBACK_M above -- the server sends null rather than a number invented there, because
 * an invented one would arrive indistinguishable from a measurement. Four metres is half a
 * foundation tile, which is small enough not to overstate an uploader and big enough to be
 * clickable at the zoom the layer is meant to be read at. */
var STORAGE_FALLBACK_M = 4;

/* What is in one container, as popup rows.
 *
 * TWO KINDS, AND ONLY ONE OF THEM GETS A GRID. A solid container holds stacks of things you
 * would recognise by their pictures, which is exactly what `contentsRows` draws; a fluid
 * buffer holds ONE fluid and a level, and a grid of a single tile would be a picture claiming
 * to be a set. Worse, the tile has nowhere to put the reading: what a reader wants off a tank
 * is "1,441 m³ of 2,400 — 60% full", and a corner badge cannot say a denominator. So the
 * fluid branch stays the sentence it always was, and the split is on `kind`, which is what
 * the server says to branch on.
 */
function storageContents(s: StorageRow): Row[] {
  if (s.kind === "fluid") {
    var stored = s.stored_m3;
    if (stored === null || stored === undefined) return [["contents", "not recorded"]];
    var level = count(Math.round(stored * 10) / 10) + " m³";
    // The capacity is what turns a level into a reading, and it comes from the docs dump
    // rather than the save -- so where the dump is silent the row says the level alone
    // instead of inventing a denominator.
    if (s.capacity_m3) {
      level += " of " + count(s.capacity_m3) + " — " + Math.round((s.fill || 0) * 100) + "% full";
    }
    return [
      ["fluid", s.fluid_name || (s.fluid ? null : "empty")],
      ["level", level],
    ];
  }
  // ...and a solid container is an inventory, so it is drawn as one -- the same grid a crate
  // gets, out of the same helper, because "what is in it" is one question wherever it is
  // asked. The truncation is the server's: `/api/storage` sends the biggest six kinds and
  // counts the rest, and `more` is what the grid's last tile stands for.
  return contentsRows(s.items || [], s.more || 0);
}

/* One container's whole card: what it is, what is in it, and where it stands.
 *
 * The contents come FIRST, above the placement rows every other popup on this page leads with,
 * because they are the reason this layer exists -- a reader who clicks a box is asking what is
 * in it, not where it is, and where it is was answered by the click.
 */
function storagePopup(s: StorageRow): Row[] {
  var rows: Row[] = [["storage", s.name]];
  storageContents(s).forEach(function (row) {
    rows.push(row);
  });
  rows.push(["slots", s.slots ? s.slots + " slots" : null]);
  rows.push(["footprint", s.w_m && s.l_m ? s.w_m + " x " + s.l_m + " m" : null]);
  rows.push(["facing", s.yaw === null || s.yaw === undefined ? null : Math.round(s.yaw) + "°"]);
  rows.push(["at", s.x_m + ", " + s.y_m + " m"]);
  rows.push(["instance", code(s.instance_leaf)]);
  return rows;
}

export function drawStorage(data: StorageResponse): void {
  // Off at the whole-world zoom, exactly like the machines and the routes: 151 boxes across
  // 7 km is a scatter of specks, and the owner asked for a toggle.
  // Last of the built band: a container is the thing a base is built AROUND, and the row is
  // off by default, so the bottom of the list is where a reader who wants it goes looking.
  var group = layer("storage", false, STORAGE_COLOUR, [BAND.built, 70, "storage"]);
  data.storage.forEach(function (s) {
    if (s.x_m === null || s.y_m === null) return;
    var colour = s.kind === "fluid" ? STORAGE_FLUID_COLOUR : STORAGE_COLOUR;
    var w = (s.w_m || STORAGE_FALLBACK_M) / 2;
    var l = (s.l_m || STORAGE_FALLBACK_M) / 2;
    var box = L.polygon(footprintCorners(s.x_m, s.y_m, w, l, s.yaw), {
      color: colour,
      weight: 1,
      fillColor: colour,
      // A container the player has not filled is drawn hollow, which is the one thing about a
      // warehouse a reader wants at a glance and the map can say without being asked: an empty
      // box is a place with room in it. Same device the machines use for `paused`.
      fillOpacity: s.kind === "fluid" ? (s.fill ? 0.7 : 0.15) : s.total ? 0.7 : 0.15,
    })
      // Wider than the page's other cards, because this one lists item names against counts
      // and a name is not broken across lines. See CONTENTS_POPUP_PX in dom.ts.
      .bindPopup(popup(storagePopup(s)), { maxWidth: CONTENTS_POPUP_PX });
    // WHERE it stands, and deliberately no instance id: `/api/floors` does not decompose
    // storage, so there is no band listing this box and a mark carrying an id would be a
    // join that always misses. Position is the honest one, and floors.ts says so.
    box._floor = { x_m: s.x_m, y_m: s.y_m, z_m: s.z_m === null ? undefined : s.z_m };
    box.addTo(group);
  });
  raiseNodeDots();
}

/* Static rather than live, which is a claim about the CONTAINER and not about its contents:
 * the boxes move when the player builds. What is in them changes on every autosave and is not
 * refetched until the next switch, which is the same bargain the machines' clock speeds make
 * -- and the reason it is bearable is that the popup is opened by a click, on demand. */
registerFetch<StorageResponse>({
  wave: "static",
  rank: 60,
  path: "/api/storage",
  label: "storage",
  clears: ["storage"],
  refilters: true,
  draw: drawStorage,
});

File: src/satisfactory_mcp/interfaces/web/routers/placements.py
"""``/api/machines`` and ``/api/structures``: everything the player physically placed.

Two endpoints and one row builder, in one file because they answer one question in two
resolutions. A machine is an ACTOR -- it has an instance id, a recipe, a clock -- and a
foundation is a lightweight buildable with none of those, interned into a positional table
because a record per piece would be megabytes. What they share is that both are things
standing somewhere with a size and a facing, which is the whole of what a map draws.

WARNING: the function name is the operation_id -- rename it and the committed schema
churns. FastAPI's default id is ``{function_name}_{path}_{method}`` and ``api-schema.d.ts``
is generated off it.

**Declaration order is wire order** for the TypedDicts below, and a ``response_model``
FILTERS -- both rules are written out at length on ``FloorsResponse`` in routers/floors.py
and are not restated per file.

THE ONE THING THIS FILE DECIDES that its neighbours do not: an ACTOR record always carries
a class, and an INTERNED table row may not. ``machines``/``extractors``/``generators`` are
actor records, written by ``saveio.extract`` behind ``if not cls.startswith("Build_")``, so
``cls`` is a non-empty string on every one of them and ``building_name`` therefore never
falls through to a ``None``. ``structures`` is the interned table, where a class is an INDEX
into a legend and ``saveio.rows`` types ``Structure.cls`` as ``str | None`` for the row whose
index points past the end -- a real piece at a real place with no name. So the two row shapes
below differ on ``cls``/``name`` for a reason, and it is not an oversight.
"""

from __future__ import annotations

from typing import Any, TypedDict

from fastapi import APIRouter, Request

from ....core.gamedata.footprint import FOUNDATION_M
from ....core.gamedata.model import pretty_class
from ....core.saveio import rows as saverows
from ....domain.world.state import WorldState
from ..serial import _fail, _m, _state, _xyz, _yaw

__all__ = ["router"]

router = APIRouter(prefix="/api")


# --------------------------------------------------------------------- helpers


class PlacementRow(TypedDict):
    """A machine, an extractor or a generator: one row shape, three layers.

    ``cls`` and ``name`` are NOT nullable, which is a claim about these three keys of the
    projection rather than about ``_record_row``. They are actor records and every one of
    them was written past ``cls.startswith("Build_")``, so the id is there and
    ``building_name`` resolves it or renders it -- see the module docstring for the line
    this draws against ``/api/structures`` next door.

    Everything a coordinate helper touches IS nullable, and each for its own reason.
    ``x_m``/``y_m``/``z_m``: an actor whose transform did not decode has no ``pos`` and
    ``_xyz`` answers with a triple of nulls -- the page's ``drawPlacements`` skips on
    exactly that. ``yaw``: ``null`` means the projection predates schema 12 and the facing
    was never recorded, which is a different claim from a facing of zero.

    ``clock`` is ``float | None`` and the float is load-bearing: ``mCurrentPotential``
    reaches the projection through ``round(float(...), 6)``, so 250% is ``2.5`` and 200% is
    ``2.0`` -- an ``int`` here would reject the first and a machine with no overclock
    property at all sends no ``clock`` key, which is the null. ``recipe_name`` is null on
    the same terms ``pretty_class`` is: no recipe in, no words out.

    ``w_m``/``l_m``/``h_m`` go null TOGETHER -- one clearance box, read whole or not at all
    -- for the 470 of 539 buildings the docs dump carries no ``mClearanceData`` for. All
    three are floats where they are anything: ``Footprint`` is metres already and this
    layer only rounds.
    """

    instance_leaf: str
    cls: str
    name: str
    x_m: float | None
    y_m: float | None
    z_m: float | None
    recipe: str | None
    recipe_name: str | None
    clock: float | None
    paused: bool
    yaw: float | None
    w_m: float | None
    l_m: float | None
    h_m: float | None


class MachinesResponse(TypedDict):
    """What ``/api/machines`` sends on a 200. An error is a 4xx with ``{"error": ...}``.

    Three keys and no counts, which is the payload's own shape: the handler is a dict
    comprehension over the three projection keys and the page reads ``data[kind]``.
    """

    machines: list[PlacementRow]
    extractors: list[PlacementRow]
    generators: list[PlacementRow]


class StructureRow(TypedDict):
    """One lightweight buildable: a foundation, a ramp, a wall, a catwalk.

    ``cls`` is nullable HERE and not on ``PlacementRow``, and the module docstring says why:
    this is the interned table, a class is an index into a legend, and a row whose index
    points past the end is a real piece at a real place with no name. ``saveio.rows`` types
    it that way and this endpoint passes ``piece.cls`` straight through.

    The three coordinates are NOT nullable, and that is ``iter_structures``' own refusal
    rather than this layer's: a row whose x, y or z will not read as a number is DROPPED
    there, so a piece that reaches ``_m`` here has all three and ``_m`` of a float is a
    float. This is the same shape of claim ``/api/nodes`` makes about its own triple --
    ``_xyz`` is broad because it also serves placements, whose transforms can fail, and
    neither of those two endpoints has a transform to fail.

    ``yaw`` is the one that survives being unreadable: ``null`` for a schema-11 row with no
    fifth column at all, and for the schema-16 rotation that will not decode.
    """

    cls: str | None
    x_m: float
    y_m: float
    z_m: float
    yaw: float | None


class StructuresResponse(TypedDict):
    """What ``/api/structures`` sends on a 200. An error is a 4xx with ``{"error": ...}``.

    ``tile_m`` is a float because ``FOUNDATION_M`` is ``8.0``: declaring it ``int`` would
    validate 8.0 into 8 and rewrite the bytes on the wire.
    """

    structures: list[StructureRow]
    count: int
    tile_m: float


def _record_row(st: WorldState, row: dict) -> PlacementRow:
    """One machine/extractor/generator, flattened for the map.

    ``clock`` and ``paused`` are read with ``get``: the projection only carries them
    for the records that have them, and an extractor at 250% and a constructor with no
    overclock property must both come out of here without a KeyError.

    ``w_m``/``l_m`` are the building's own footprint -- the X and Y extent of the union
    of its clearance boxes, which is what makes a Manufacturer draw bigger than a
    Constructor instead of both being the same nominal square. Only 69 of 539 buildings
    carry clearance data, so these are **null** for the rest (the two biomass burners on
    the reference save among them) rather than a guessed number: the client picks the
    fallback, because a fallback drawn here would be indistinguishable from a measurement.

    ``yaw`` is which way the building faces, and it is what turns ``w_m``/``l_m`` from an
    axis-aligned box into the rectangle the player actually placed -- the two are one
    answer and are read together or not at all.

    ``h_m`` is the third side of the same box, and it is the one dimension a top-down map
    cannot show: it is here because a FLOOR view needs it. A Refinery is 15 m tall standing
    on a 12 m storey, so it comes three metres through the deck above and is physically in
    the way of anything built there -- a fact about the floor above that can only be read
    off the floor below. Null on exactly the same terms as ``w_m``/``l_m``, from the same
    ``mClearanceData``, and a client that has no height draws no such claim.
    """
    cls = row.get("cls") or ""
    building = st.game.buildings.get(cls)
    footprint = getattr(building, "footprint", None) if building else None
    recipe_id = row.get("recipe")
    recipe = st.game.recipes.get(recipe_id) if recipe_id else None
    return {
        "instance_leaf": str(row.get("instance", "")).rsplit(".", 1)[-1],
        "cls": row.get("cls"),
        # The docs name where the dump has one; readable words either way. The raw class
        # stays in ``cls`` for anything that needs the exact id.
        "name": st.game.building_name(cls),
        **_xyz(row.get("pos")),
        "recipe": recipe_id,
        # Same rule for the recipe: the row a player most wants to read must not be the
        # one row still speaking engine ids.
        "recipe_name": recipe.name if recipe else pretty_class(recipe_id),
        "clock": row.get("clock"),
        "paused": bool(row.get("paused", False)),
        "yaw": _yaw(row.get("yaw")),
        # Footprint is already metres; the projection's coordinates are not.
        "w_m": round(footprint.width_m, 1) if footprint else None,
        "l_m": round(footprint.depth_m, 1) if footprint else None,
        "h_m": round(footprint.height_m, 1) if footprint else None,
    }


# ------------------------------------------------------------------- machines


@router.get("/machines", response_model=MachinesResponse)
def machines(request: Request, save: str | None = None, world: str | None = None) -> Any:
    try:
        st = _state(request, save, world)
    except Exception as exc:
        return _fail(f"could not read save: {exc}", 404)
    p = st.projection
    return {
        kind: [_record_row(st, row) for row in p.get(kind, ())]
        for kind in ("machines", "extractors", "generators")
    }


# ----------------------------------------------------------------- structures


@router.get("/structures", response_model=StructuresResponse)
def structures(request: Request, save: str | None = None, world: str | None = None) -> Any:
    """Every lightweight buildable the player placed: foundations, ramps, walls, catwalks.

    These are the only record of what was physically BUILT -- they appear in no actor
    header, which is why the projection interns them separately as
    ``{"classes": [...], "instances": [[class_index, x, y, z, yaw], ...]}`` in centimetres.
    Decoded by ``core.saveio.rows``, which is where the guard lives for all ten readers
    of these interned tables: a malformed row costs one piece, not the endpoint.

    **Rotation is carried, as of schema 12**, and this docstring used to say the opposite
    -- the instance transform's quaternion was dropped at extraction and a client could
    only draw these axis-aligned, which is why an angled slab came out of the map as a
    staircase of squares. ``yaw`` is now the fifth column of a row and comes out as a
    ``yaw`` field: degrees about world Z, positive turning +X towards +Y. ``null`` for a
    projection cut before 12, which a client must keep drawing axis-aligned rather than
    reading as zero.

    One thing the projection still does not carry, and it is not invented here:

    * **Per-class size.** None of these classes has clearance data, so ``footprint`` is
      ``None`` for all eighteen of them. They are all built on the same grid instead,
      whose edge ``tile_m`` reports from ``FOUNDATION_M`` so the page does not hardcode 8.

    Positions are piece centres: on the reference save consecutive foundations of one
    slab sit exactly ``tile_m`` apart.

    A world with nothing built answers ``{"structures": [], "count": 0}`` -- an empty
    list is a real answer here, unlike a save that could not be read at all.

    Sent one row per piece, ungrouped. Measured on the reference world -- 8,347 pieces,
    708 KB (610 KB of it before the yaw column) -- which is the same order as
    ``/api/collectibles`` already ships (3,455 rows, 547 KB). Grouping into grid cells would halve a
    payload that is not the bottleneck and would cost the per-piece class the popup and
    the point inspector read.
    """
    try:
        st = _state(request, save, world)
    except Exception as exc:
        return _fail(f"could not read save: {exc}", 404)

    out = [
        {
            "cls": piece.cls,
            "x_m": _m(piece.x),
            "y_m": _m(piece.y),
            "z_m": _m(piece.z),
            # Optional on purpose: a row from a schema-11 projection is four columns long
            # and is still a real piece at a real place, it just has no facing. ``None``
            # covers the schema-16 unreadable rotation too -- see ``saveio.rows``.
            "yaw": _yaw(piece.yaw),
        }
        for piece in saverows.iter_structures(st.projection)
    ]
    return {"structures": out, "count": len(out), "tile_m": FOUNDATION_M}
